### ISSUE TITLE
JobExecutionError without relevant error message

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-server#451]
I'm trying to run a new example and my workflow fails during the `Convert Yokogawa to OME-Zarr` with a `JobExecutionError`.

The weird thing is that the .err file of the job says the job finished successfully:
```
Matplotlib created a temporary config/cache directory at /tmp/matplotlib-wrbssjh0 because the default path (/data/homes/fractal/.cache/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
2023-01-16 13:49:38,878; INFO; START yokogawa_to_ome_zarr task
2023-01-16 13:49:39,263; INFO; glob path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*_B03_*A01*C01*.png
2023-01-16 13:49:55,731; INFO; glob path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*_B03_*A01*C02*.png
2023-01-16 13:50:09,745; INFO; glob path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*_B03_*A02*C03*.png
2023-01-16 13:50:39,890; INFO; END yokogawa_to_ome_zarr task
```

And the slurm .err file only contains this:
```
Content of /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.slurm.9445845.err:
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/0.slurm.submit.sbatch': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.slurm.submit.sbatch': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/cfut.in.uoaLSyH2xsDzvmmasowgIC7ahBxNWwbF.pickle': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/workflow.log': Operation not permitted
```

But that also appears in jobs that don't fail.

Thus, I don't understand why this job is failing (currently looking into this) and it's worrying that the job fails without a clear error message...
@tcompa the chmod failing can't actually fail the job, right? Otherwise, I wouldn't understand why this reproducibly happens in the workflow here, but not in other examples...

I'll check with a different executor with more memory (shouldn't actually be an issue, but who knows) and a different example script.

**Comment by tcompa** [comment]
> I'm trying to run a new example and my workflow fails during the `Convert Yokogawa to OME-Zarr` with a `JobExecutionError`.
> 
> The weird thing is that the .err file of the job says the job finished successfully:
> 
> ```
> Matplotlib created a temporary config/cache directory at /tmp/matplotlib-wrbssjh0 because the default path (/data/homes/fractal/.cache/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
> 2023-01-16 13:49:38,878; INFO; START yokogawa_to_ome_zarr task
> 2023-01-16 13:49:39,263; INFO; glob path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*_B03_*A01*C01*.png
> 2023-01-16 13:49:55,731; INFO; glob path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*_B03_*A01*C02*.png
> 2023-01-16 13:50:09,745; INFO; glob path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*_B03_*A02*C03*.png
> 2023-01-16 13:50:39,890; INFO; END yokogawa_to_ome_zarr task
> ```
> 
> And the slurm .err file only contains this:
> 
> ```
> Content of /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.slurm.9445845.err:
> chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/0.slurm.submit.sbatch': Operation not permitted
> chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.slurm.submit.sbatch': Operation not permitted
> chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/cfut.in.uoaLSyH2xsDzvmmasowgIC7ahBxNWwbF.pickle': Operation not permitted
> chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000006_job_000006/workflow.log': Operation not permitted
> ```
> 
> But that also appears in jobs that don't fail.
> 
> Thus, I don't understand why this job is failing (currently looking into this) and it's worrying that the job fails without a clear error message... @tcompa the chmod failing can't actually fail the job, right? Otherwise, I wouldn't understand why this reproducibly happens in the workflow here, but not in other examples...
> 
> I'll check with a different executor with more memory (shouldn't actually be an issue, but who knows) and a different example script.

chmod fails for a known reason (it's a blanket application of chmod on all files in the folder, including those that belong to fractal). We never cleaned it up, since we plan to replace it - ref https://github.com/fractal-analytics-platform/fractal-server/issues/341

As for the rest: agreed, it's weird that nothing else shows up.
Is this with SLURM or process backend?
Can you provide the log (the one where you saw the JobExecutionError) or at least some additional information?

**Comment by jluethi** [comment]
This is with SLURM process.

When I change to a cpu-mid executor, it does not fail. Almost as if it ran out of memory after finishing the actual content of the job, for some reason... Running more tests atm. Have we tested how out of memory errors would be caught?

The logs above are the full content of the two .err files. The workflow log contains this. Any other logs you'd be interested in?
```
2023-01-16 10:43:16,128; INFO; fractal_server.__VERSION__: 1.0.0rc1
2023-01-16 10:43:16,129; INFO; FRACTAL_RUNNER_BACKEND: slurm
2023-01-16 10:43:16,129; INFO; slurm_user: jluethi
2023-01-16 10:43:16,129; INFO; worker_init: export CELLPOSE_LOCAL_MODELS_PATH=/data/homes/jluethi/.cache/CELLPOSE_LOCAL_MODELS_PATH
export NUMBA_CACHE_DIR=/data/homes/jluethi/.cache/NUMBA_CACHE_DIR
export NAPARI_CONFIG=/data/homes/jluethi/.cache/NAPARI_CACHE_DIR
export XDG_CONFIG_HOME=/data/homes/jluethi/.cache/XDG

2023-01-16 10:43:16,129; INFO; input metadata: {}
2023-01-16 10:43:16,129; INFO; input_paths: [PosixPath('/data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/../images/10.5281_zenodo.7057076/*.png')]
2023-01-16 10:43:16,129; INFO; output_path: /data/homes/jluethi/v1_fractal/fractal-demos/examples/02_cardio_small/tmp_cardio-2x2-1/output/*.zarr
2023-01-16 10:43:16,139; INFO; job.id: 5
2023-01-16 10:43:16,139; INFO; job.working_dir: /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000005_job_000005
2023-01-16 10:43:16,139; INFO; START workflow "Workflow cardio-2x2-1"
2023-01-16 10:43:16,140; INFO; START 0-th task (name="Create OME-Zarr structure", executor=cpu-low).
2023-01-16 10:43:16,514; INFO; END   0-th task (name="Create OME-Zarr structure", executor=cpu-low).
2023-01-16 10:43:16,515; INFO; START 1-th task (name="Convert Yokogawa to OME-Zarr", executor=cpu-low).
2023-01-16 10:46:02,366; INFO; FAILED workflow "Workflow cardio-2x2-1", JobExecutionError.
```

**Comment by jluethi** [comment]
I now pushed the example that reproduces this for me to the fractal-demos repo here: https://github.com/fractal-analytics-platform/fractal-demos/tree/main/examples/02_cardio_small

In this (and in other examples when using the same input data), "Convert Yokogawa to OME-Zarr" always fails on its default executor on the slurm setup.
Using the cpu-mid executor (as in the currently committed code) solves this issue. But I can't trace why it fails on the cpu-low executor.

The output (even of a "failed" run) is the correctly parsed OME-Zarr file (just without the pipeline continuing). And this task has not given us memory issues in the past. But even if it were: If this was e.g. a memory issue, we'd expect to see something about this in the log, no?

**Comment by tcompa** [comment]
> Have we tested how out of memory errors would be caught?

One would guess that they fall under the general case of SLURM errors, but we've not tested that case specifically (AKA we don't have it in the CI tests).

> The logs above are the full content of the two .err files. The workflow log contains this. Any other logs you'd be interested in?

Is there anything more in the output of `fractal job show 5`?

**Comment by tcompa** [comment]
Is this at FMI or UZH?

At FMI if SLURM kills the job because of memory limits I would expect to see something in the SLURM stderr file (as in https://github.com/fractal-analytics-platform/fractal-server/issues/343#issue-1496330932)

**Comment by jluethi** [comment]
It's at UZH

**Comment by jluethi** [comment]
I'll run the same test at FMI to check :)

**Comment by tcompa** [comment]
> Do we have a way to check memory usage of a slurm job? e.g. through slurm itself somehow?

You can play around with commands like
```
sacct --format='JobID,Elapsed,MaxRSS,MaxVMSize,ExitCode' -u YOUR_USER -b
```

Or (at FMI) iteratively change the memory limit and see when things start failing.

**Comment by tcompa** [comment]
Actually the memory-error handling can be tested directly in a case where we know that this is the issue, see https://github.com/fractal-analytics-platform/fractal-server/issues/343#issue-1496330932 (note that this would be a test to be done at FMI, since behavior at UZH would be different).

**Comment by jluethi** [comment]
Hmm, this is weird. The .0 job always seems to complete, while the .bat+ job fails. Don't really understand enough about slurm to get the different between the .0 and the .bat+ job.

![Screenshot 2023-01-16 at 15 29 41](https://user-images.githubusercontent.com/18033446/212701762-5d590a3e-f825-467e-bb5c-a72a883ea6e6.png)

I'll run the test at FMI to see how they behave there.

**Comment by tcompa** [comment]
With `--format='JobID,JobName,Elapsed,MaxRSS,MaxVMSize,ExitCode'` you'll notice that part of it is the batch script, and the other part is a python script (i.e. the actual task).

Also note that the FAILED SLURM statuses are likely due to the chmod issue, rather than to some actual failure.

**Comment by jluethi** [comment]
Ok... The mystery deepens:
I can reproducibly make it fail this way on the UZH pelkmans lab slurm cluster. But it runs fine on the FMI test slurm cluster... Super suspicious that it fails without a clearer error message though. But then it's likely not a memory thing, because that would also occur on the FMI side (and actually fail correctly over there).

**Comment by jluethi** [comment]
To keep track of things happening:
Example 02 also fails the same way at the "Maximum Intensity Projection" task on the UZH setup. Retrying now with a cpu-mid executor. Still no idea why it fails though

**Comment by jluethi** [comment]
Brief update: it even fails with cpu-mid during "Maximum Intensity Projection":


```
  "history": [
    "Create OME-Zarr structure",
    "Convert Yokogawa to OME-Zarr: ['20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/']",
    "Illumination correction: ['20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/']",
    "Copy OME-Zarr structure"
  ]
}

This is the job log:

JOB ERROR:
TRACEBACK:
JobExecutionError

COMMAND:
Content of /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/3.slurm.submit.sbatch:
#!/bin/sh
#SBATCH --output=/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/3.slurm.%j.out
#SBATCH --error=/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/3.slurm.%j.err
#SBATCH --partition=main
#SBATCH --mem=15g
#SBATCH --cpus-per-task=4
#SBATCH --job-name Copy_OME-Zarr_structure
export SRUN_CPUS_PER_TASK=4
export CELLPOSE_LOCAL_MODELS_PATH=/data/homes/jluethi/.cache/CELLPOSE_LOCAL_MODELS_PATH
export NUMBA_CACHE_DIR=/data/homes/jluethi/.cache/NUMBA_CACHE_DIR
export NAPARI_CONFIG=/data/homes/jluethi/.cache/NAPARI_CACHE_DIR
export XDG_CONFIG_HOME=/data/homes/jluethi/.cache/XDG

export CFUT_DIR=/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012
srun /data/homes/fractal/.conda/envs/fractal-server-1.0.0rc1/bin/python -m fractal_server.app.runner._slurm.remote /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/cfut.in.jI9iy8elq9JCY7QEMZPxjJRPc90HvzEG.pickle
chmod 777 /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/*

STDOUT:
File /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/3.slurm.9445865.out is empty


STDERR:
Content of /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/3.slurm.9445865.err:
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/0.slurm.submit.sbatch': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.slurm.submit.sbatch': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/2_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.slurm.submit.sbatch': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/3.slurm.submit.sbatch': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/cfut.in.jI9iy8elq9JCY7QEMZPxjJRPc90HvzEG.pickle': Operation not permitted
chmod: changing permissions of '/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples/server/artifacts/workflow_000012_job_000012/workflow.log': Operation not permitted
```

**Comment by jluethi** [comment]
Just to make sure: The chmod errors can't be the source of this, right? I see them even when jobs pass, so seems very unlikely...

**Comment by tcompa** [comment]
The behavior is different on the `process` backend.
The task fails with a `TaskExecutionError` during the yokogawa-to-zarr task (the second one in the list, after the create-ome-zarr one). This leads to (at least) two obvious questions:
1. Why does the task fail? (see upcoming issue in the tasks repo)
2. Why does the server "sometimes" see a `JobExecutionError`?

Here are other (less precise) info:

3. While playing around with the SLURM backend, on the UZH cluster, I also managed to observe the same `TaskExecutionError` as on the `process` backend. I think it was when I removed the `chmod` final line from the sbatch script, but I should re-check more carefully.
4. While playing around with the SLURM backend, on the UZH cluster, I also managed to observe the same `JobExecutionError` as @jluethi did.

**Comment by jluethi** [comment]
Ok, very good to know. 

> While playing around with the SLURM backend, on the UZH cluster, I also managed to observe the same TaskExecutionError as on the process backend. I think it was when I removed the chmod final line from the sbatch script, but I should re-check more carefully.

Is it possible that it's always a TaskError, but the chmod errors made Fractal think it's a `JobExecutionError`? Would be my major suspicion here...

**Comment by tcompa** [comment]
As a side comment, this also explains why the zarr files exist and are OK. The task actually got to the end, but something went wrong upon writing the actual output pickle file (which is a file used internally by the executor, and has nothing to do with zarr).

**Comment by tcompa** [comment]
For the record, the task that fails with this error is not fixed.

I have for sure observed "Convert Yokogawa to OME-Zarr" failing in previous runs, and [also the MIP task failed](https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/283).
Now I have a ongoing workflow run that is at this state:
```
2023-01-19 11:09:21,772; INFO; input metadata: {}
2023-01-19 11:09:21,772; INFO; input_paths: [PosixPath('/data/homes/test01/fractal-demos/examples/02_cardio_small-debug/../images/10.5281_zenodo.7057076/*.png')]
2023-01-19 11:09:21,772; INFO; output_path: /data/homes/test01/fractal-demos/examples/02_cardio_small-debug/tmp_cardio-2x2-3/output/*.zarr
2023-01-19 11:09:21,784; INFO; job.id: 1
2023-01-19 11:09:21,784; INFO; job.working_dir: /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/fractal-server/debug/artifacts/workflow_000001_job_000001
2023-01-19 11:09:21,784; INFO; START workflow "Workflow cardio-2x2-3"
2023-01-19 11:09:21,786; INFO; START 0-th task (name="Create OME-Zarr structure", executor=cpu-low).
2023-01-19 11:09:22,155; INFO; END   0-th task (name="Create OME-Zarr structure", executor=cpu-low).
2023-01-19 11:09:22,155; INFO; START 1-th task (name="Convert Yokogawa to OME-Zarr", executor=cpu-mid).
2023-01-19 11:10:52,922; INFO; END   1-th task (name="Convert Yokogawa to OME-Zarr", executor=cpu-mid).
2023-01-19 11:10:52,923; INFO; START 2-th task (name="Illumination correction", executor=cpu-low).
2023-01-19 11:11:54,011; INFO; END   2-th task (name="Illumination correction", executor=cpu-low).
2023-01-19 11:11:54,011; INFO; START 3-th task (name="Cellpose Segmentation", executor=gpu).
2023-01-19 11:13:55,183; INFO; END   3-th task (name="Cellpose Segmentation", executor=gpu).
2023-01-19 11:13:55,183; INFO; START 4-th task (name="Napari workflows wrapper", executor=cpu-low).
2023-01-19 11:16:26,400; INFO; END   4-th task (name="Napari workflows wrapper", executor=cpu-low).
2023-01-19 11:16:26,401; INFO; START 5-th task (name="Copy OME-Zarr structure", executor=cpu-low).
2023-01-19 11:16:26,766; INFO; END   5-th task (name="Copy OME-Zarr structure", executor=cpu-low).
2023-01-19 11:16:26,767; INFO; START 6-th task (name="Maximum Intensity Projection", executor=cpu-low).
INFO:WF1_job1:FAILED workflow "Workflow cardio-2x2-3", JobExecutionError.
```
And nothing changed in the server code, apart from some more detailed logs (which also include some `time.sleep` statements, but that's in a part of the code that was not executed in this successful case).

**Comment by tcompa** [comment]
Anecdotal information: I saw a spike in SLURM usage (by other users) close to the failure time. It may be a coincidence, but let's not exclude the case of nfs4 file system becoming extremely slow for a short period of time.

If writing the pickle file to disk takes even a couple of seconds, then the server has already entered the output-pickle-file-does-not-exist branch, and will fail.

The following log clearly points to this direction:
```
ADDITIONAL INFO:
Output pickle file /net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/fractal-server/debug/artifacts/workflow_000001_job_000001/cfut.out.Hku4JzGtU5J0iCVvx1vVFkk8DYevW0Ex.pickle not found.

Reality check:
out_path.exists()=False
os.path.exists(str(out_path))=False

-- now sleep 5 seconds --

out_path.exists()=True
os.path.exists(str(out_path))=True

-- now sleep 5 seconds --

out_path.exists()=True
os.path.exists(str(out_path))=True

-- now sleep 5 seconds --

out_path.exists()=True
os.path.exists(str(out_path))=True
```

Quick&dirty fix:
Give a grace period to the file-not-found code branch. If it does not exist, wait 5 seconds and then re-try. If it still does not exist, fail with a JobExecutionError.

**Comment by tcompa** [comment]
TL;DR
We implemented a workaround which made the test work. I don't see any further actions required, apart from more testing.

---

The first test using [this branch](https://github.com/fractal-analytics-platform/fractal-server/tree/451-jobexecutionerror-without-relevant-error-message) (as of [this commit](https://github.com/fractal-analytics-platform/fractal-server/pull/459/commits/f39fa6894186753d56aa8fd2d4c459f07a445685)) ran through correctly, with the default
```
FRACTAL_SLURM_OUTPUT_FILE_GRACE_TIME = 4
```
And the status is:
```
$ fractal job show 1
{
  "project_id": 1,
  "input_dataset_id": 1,
  "output_dataset_id": 2,
  "workflow_id": 1,
  "overwrite_input": false,
  "worker_init": "export CELLPOSE_LOCAL_MODELS_PATH=/data/homes/test01/.cache/CELLPOSE_LOCAL_MODELS_PATH\nexport NUMBA_CACHE_DIR=/data/homes/test01/.cache/NUMBA_CACHE_DIR\nexport NAPARI_CONFIG=/data/homes/test01/.cache/NAPARI_CACHE_DIR\nexport XDG_CONFIG_HOME=/data/homes/test01/.cache/XDG\n",
  "working_dir": "/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/fractal-server/debug/artifacts/workflow_000001_job_000001",
  "id": 1,
  "start_timestamp": "2023-01-19 12:40:25.710277",
  "status": "done",
  "log": null,
  "history": [
    "Create OME-Zarr structure",
    "Convert Yokogawa to OME-Zarr: ['20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/']",
    "Illumination correction: ['20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/']",
    "Cellpose Segmentation: ['20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/']",
    "Napari workflows wrapper: ['20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/']",
    "Copy OME-Zarr structure",
    "Maximum Intensity Projection: ['20200812-CardiomyocyteDifferentiation14-Cycle1_mip.zarr/B/03/0/']",
    "Cellpose Segmentation: ['20200812-CardiomyocyteDifferentiation14-Cycle1_mip.zarr/B/03/0/']",
    "Napari workflows wrapper: ['20200812-CardiomyocyteDifferentiation14-Cycle1_mip.zarr/B/03/0/']",
    "Napari workflows wrapper: ['20200812-CardiomyocyteDifferentiation14-Cycle1_mip.zarr/B/03/0/']",
    "Napari workflows wrapper: ['20200812-CardiomyocyteDifferentiation14-Cycle1_mip.zarr/B/03/0/']",
    "Napari workflows wrapper: ['20200812-CardiomyocyteDifferentiation14-Cycle1_mip.zarr/B/03/0/']"
  ]
}
```

---

Concerning the current workaround:

Pros:
* It is really simple and transparent: if that pickle file is missing, just wait a little bit to be sure it's not just being written right now.

Cons:
* It adds one more env variable to be aware of. _Comment: minor issue._
* It adds 4 seconds to every failing workflow. _Comment: IMO this is also a very minor issue, at least because there are already other slow-downs related to failing workflows._
* We don't know the reference scale for this interval, at a difference with other variables (e.g. [FRACTAL_SLURM_KILLWAIT_INTERVAL](https://fractal-analytics-platform.github.io/fractal-server/configuration/#fractal_server.config.Settings.FRACTAL_SLURM_KILLWAIT_INTERVAL), which needs to be larger than a certain SLURM configuration variable). _Comment: this is the most annoying issue. We can always hope that a few seconds are enough to write a tiny file, but things can always break if suddenly this operation takes a minute. In that case, the only option would be to restart the workflow, because it's hard to trust the results if "something" went wrong in the middle._

**Comment by jluethi** [comment]
Great, thanks for digging into this. Sounds like a reasonable workaround! And with the additional logging, we're now at least getting some hints in the logs on why a workflow failed :)

Just for the bigger picture understanding: Is this the pelkmans lab SLURM being slow or the filesystem servers having issues? My default assumption is often the filesystem server, though I think (not sure though) they are user specific. In that case, a filesystem server probably shouldn't be the issue here, given that when I ran it on my share, there was not much else going on on the share (to be checked whether they really are fully independent)
